multirpc/subpub/discovery.go

```diff
@@ -79,7 +81,9 @@ func (ps *SubPub) Subscribe(ctx context.Context) {
 		case <-ps.close:
 			return
 		default:
-			ps.discover(ctx)
+			pctx, cancel := context.WithTimeout(ctx, ps.DiscoveryPeriod)
+			ps.discover(pctx)
+			cancel()
```

Comment:
I added this because we call discover in a loop, so with just the parent ctx, the ps.discover call could hang or get stuck for a long time. I assume we want to set some relatively low timeout here, like DiscoveryPeriod.